### DISCUSSION TITLE
How to display settings without clicking on element?

### DISCUSSION THREAD

**Discussion by throne1986** [prevwong/craft.js#227] [Q&A]
I mean assume I want the user to be able to change the color of all buttons in my landing page to be red.

So far user need to click on a button and after that, he can change this button setting eg color

How can I achieve this? what do I need to change to show those settings without clicking on an element?

![image](https://user-images.githubusercontent.com/27904724/112065175-7d4b2800-8b64-11eb-91f8-438380cb1942.png)

**Comment by ankri** [comment]
Interesting question. I think I would solve it like this:

- Render a `<ThemeProvider>` around the `<Editor />`. This `<ThemeProvider />` should expose a method to change colors
- In the settings panel I would display the color selection when no Component is selected
- Changing the color in the settings panel will use the methods to change colors from the `<ThemeProvider />`. The Theme Provider will change the color on its theme

---

If you ship your own Design System I would also argue to use `CSS Variables`. The components will just use the `CSS Variables` and would not access the ThemeProvider (because there is no need, since they get their color from the `CSS Variable`). Changing the colors would just change the `CSS Variables` under `:root` or a dedicated `DOM Element`. The advantage of CSS Variables would be that the components would not need to re-render when we change the colors. Re-Render in React's Context. They still need to re-draw but that is not as expensive as React's render.

**Reply by throne1986** [reply]
Thank you so much for the explanation, If you don't mind could you provide a basic demo in codesandbox.io with the option to change colors eg buttons for the whole app? for us the newbies to this great library it will be so much helpfully,

God bless your hand brother

**Comment by nicosh** [comment]
Probably @ankri solution is the most elegant, however in my case was a pain to implement such solution, since i should have to rewrite a lot of code, also im using custom function to render my components in the final product and this add even more complexity.
So i ended to use a simple function to change  props to every component:
```
    const handleBulkChange = (name, val) => {
        const data = JSON.parse(query.serialize())
        let keys = Object.keys(data)
        keys.forEach((el) => {
            actions.setProp(el, (props) => {
                props[name] = val
            })
        })
    }
```
You can also improve this function tho change props only to components with a given name : 

```
    const handleBulk = (name, val,target) => {
        const data = JSON.parse(query.serialize())
        let keys = Object.keys(data)
        keys.forEach((el) => {
            let node = data[el]
            let {displayName} = node
            if(displayName === target){
                actions.setProp(el, (props) => {
                    props[name] = val
                })
            }
        })
    }
```

**Reply by ankri** [reply]
Hi @nicosh - Thank you for your suggestion. Is there a reason why you did `JSON.parse(query.serialize())` instead of using 
`const { nodes } = useEditor((state) => ({ nodes: state.nodes })`?

**Reply by nicosh** [reply]
nope :D

**Comment by ankri** [comment]
I finally found some time to whip up an example.

[codesandbox](https://codesandbox.io/s/craft-css-vars-ib8my)

I would not recommend doing it like this though: In my example I override the color for `--chakra-colors-gray-100`. In a real application I would have used a specific `--button-background` variable. 

Also I can only recommend doing it this way, when you do not use a  DesignSystem like "material ui" or in my CodeSandbox "chakra".

**Comment by ankri** [comment]
Inspired by the answer of @nicosh I created a second [example](https://codesandbox.io/s/craft-set-all-props-h75kv). I think this would be a more ideal solution when using Chakra UI.

```ts
const { actions, nodes } = useEditor(
  (state) => {
    return { nodes: state.nodes };
  }
);
```

```ts
onChange={(event) => {
  Object.values(nodes).forEach((node) => {
    if (node.data.name === "Button") {
      actions.setProp(node.id, (props) => {
        props.colorScheme = event.target.value;
      });
    }
  });
}}
```

**Reply by nicosh** [reply]
There is only 2 things to take in account with these solutions :
1 if you want to change props to all nodes (lets say that we want change global color scheme) make sure to check that the node is not a div : 
```
    const handleBulk = (propname, val) => {
        Object.keys(nodes).forEach((el) => {
            let node = nodes[el]
            let {displayName,name} = node.data
            if(name !== "div"){
                actions.setProp(el, (props) => {
                    props[propname] = val
                })
            }         
        })
    }
```
Otherwise it will trigger an error
2 obviously it changes props only to existing nodes, new nodes will have default props.
